module/zfs/brt.c

```diff
@@ -0,0 +1,1843 @@
+	(((size) - 1) / (blocksize) / sizeof (uint64_t) + 1)
+
+#define	BRT_LITTLE_ENDIAN	0
+#define	BRT_BIG_ENDIAN		1
+#ifdef _ZFS_LITTLE_ENDIAN
+#define	BRT_NATIVE_BYTEORDER		BRT_LITTLE_ENDIAN
+#define	BRT_NON_NATIVE_BYTEORDER	BRT_BIG_ENDIAN
+#else
+#define	BRT_NATIVE_BYTEORDER		BRT_BIG_ENDIAN
+#define	BRT_NON_NATIVE_BYTEORDER	BRT_LITTLE_ENDIAN
+#endif
+
+typedef struct brt_vdev_phys {
+	uint64_t	bvp_mos_entries;
+	uint64_t	bvp_size;
+	uint64_t	bvp_byteorder;
+	uint64_t	bvp_totalcount;
+	uint64_t	bvp_rangesize;
+	uint64_t	bvp_drefsize;
+	uint64_t	bvp_dsize;
+} brt_vdev_phys_t;
+
+typedef struct brt_vdev {
+	/*
+	 * VDEV id.
+	 */
+	uint64_t	bv_vdevid;
+	/*
+	 * Is the structure initiated?
+	 * (bv_refcount and bv_bitmap are allocated?)
+	 */
+	boolean_t	bv_initiated;
+	/*
+	 * Object number in the MOS for the refcount array and brt_vdev_phys.
+	 */
+	uint64_t	bv_mos_brtvdev;
+	/*
+	 * Object number in the MOS for the entries table.
+	 */
+	uint64_t	bv_mos_entries;
+	/*
+	 * Entries to sync.
+	 */
+	avl_tree_t	bv_tree;
+	/*
+	 * Does the bv_refcount[] array needs byte swapping?
+	 */
+	boolean_t	bv_need_byteswap;
+	/*
+	 * Number of entries in the bv_refcount[] array.
+	 */
+	uint64_t	bv_size;
+	/*
+	 * This is the array with all the refcounts
+	 * (one refcount per BRT_RANGE_SIZE).
+	 */
+	uint64_t	*bv_refcount;
+	/*
+	 * Sum of all bv_refcount[]s.
+	 */
+	uint64_t	bv_totalcount;
+	/*
+	 * Disk space savings thanks to BRT.
+	 */
+	uint64_t	bv_drefsize;
+	uint64_t	bv_dsize;
+	/*
+	 * bv_refcount[] potentially can be a bit too big to sychronize it all
+	 * when we just changed few refcounts. The fields below allow us to
+	 * track updates to bv_refcount[] array since the last sync.
+	 * A single bit in the bv_bitmap represents as many refcounts as can
+	 * fit into a single brt_blocksize, where brt_blocksize is
+	 * (1 << spa->spa_min_ashift).
+	 * For example we have 65536 refcounts in the bv_refcount array
+	 * (so the whole array is 512kB). We updated bv_refcount[2] and
+	 * bv_refcount[5]. In that case only first bit in the bv_bitmap will
+	 * be set and we will write only first 4kB out of 512kB (assuming
+	 * ashift is 12).
+	 */
+	boolean_t	bv_dirty;
+	ulong_t		*bv_bitmap;
+	uint64_t	bv_nblocks;
+} brt_vdev_t;
+
+/*
+ * In-core brt
+ */
+typedef struct brt {
+	krwlock_t	brt_lock;
+	spa_t		*brt_spa;
+#define	brt_mos		brt_spa->spa_meta_objset
+	uint64_t	brt_blocksize;
+	uint64_t	brt_rangesize;
+	uint64_t	brt_drefsize;
+	uint64_t	brt_dsize;
+	avl_tree_t	brt_pending_tree[TXG_SIZE];
+	kmutex_t	brt_pending_lock[TXG_SIZE];
+	/* Sum of all entries across all bv_trees. */
+	uint64_t	brt_nentries;
+	brt_vdev_t	*brt_vdevs;
+	uint64_t	brt_nvdevs;
+} brt_t;
+
+/* Size of bre_offset / sizeof (uint64_t). */
+#define	BRT_KEY_WORDS	(1)
+
+/*
+ * In-core brt entry.
+ * On-disk we use bre_offset as the key and bre_refcount as the value.
+ */
+typedef struct brt_entry {
+	uint64_t	bre_offset;
+	uint64_t	bre_refcount;
+	avl_node_t	bre_node;
```

Comment:
avl_node_t is 24 bytes in size, that gives 40% memory efficiency.  I wonder whether B-tree with its average 75% would be better here instead, or considering those are deleted at each TXG sync it is not an issue.  Any way this is only in RAM, so can be investigated later.

Reply:
I'll leave it for now as-is. Thanks.